src/draw/dma2d/lv_draw_dma2d.c

```diff
@@ -0,0 +1,569 @@
+/**
+ * @file lv_draw_dma2d.c
+ *
+ */
+
+/*********************
+ *      INCLUDES
+ *********************/
+
+#include "lv_draw_dma2d.h"
+#if LV_USE_DRAW_DMA2D
+
+#include "../lv_draw_private.h"
+#include "../sw/lv_draw_sw.h"
+#include "../../misc/lv_area_private.h"
+
+#include "stm32h7xx_hal.h"
+#include "stm32h7xx_hal_dma2d.h"
+#include "stm32h7b3i_discovery_lcd.h"
+
+/*********************
+ *      DEFINES
+ *********************/
+
+#define DRAW_UNIT_ID_DMA2D 5
+
+/**********************
+ *      TYPEDEFS
+ **********************/
+
+typedef enum {
+    LV_DRAW_DMA2D_OUTPUT_CF_ARGB8888 = 0,
+    LV_DRAW_DMA2D_OUTPUT_CF_RGB888,
+    LV_DRAW_DMA2D_OUTPUT_CF_RGB565,
+    LV_DRAW_DMA2D_OUTPUT_CF_ARGB1555,
+    LV_DRAW_DMA2D_OUTPUT_CF_ARGB4444
+} lv_draw_dma2d_output_cf_t;
+
+typedef enum {
+    LV_DRAW_DMA2D_FGBG_CF_ARGB8888 = 0,
+    LV_DRAW_DMA2D_FGBG_CF_RGB888,
+    LV_DRAW_DMA2D_FGBG_CF_RGB565,
+    LV_DRAW_DMA2D_FGBG_CF_ARGB1555,
+    LV_DRAW_DMA2D_FGBG_CF_ARGB4444,
+    LV_DRAW_DMA2D_FGBG_CF_L8,
+    LV_DRAW_DMA2D_FGBG_CF_AL44,
+    LV_DRAW_DMA2D_FGBG_CF_AL88,
+    LV_DRAW_DMA2D_FGBG_CF_L4,
+    LV_DRAW_DMA2D_FGBG_CF_A8,
+    LV_DRAW_DMA2D_FGBG_CF_A4,
+    LV_DRAW_DMA2D_FGBG_CF_YCBCR
+} lv_draw_dma2d_fgbg_cf_t;
+
+typedef enum {
+    LV_DRAW_DMA2D_MODE_MEMORY_TO_MEMORY = 0,
+    LV_DRAW_DMA2D_MODE_MEMORY_TO_MEMORY_WITH_PFC,
+    LV_DRAW_DMA2D_MODE_MEMORY_TO_MEMORY_WITH_BLENDING,
+    LV_DRAW_DMA2D_MODE_REGISTER_TO_MEMORY,
+    LV_DRAW_DMA2D_MODE_MEMORY_TO_MEMORY_WITH_BLENDING_AND_FIXED_COLOR_FG,
+    LV_DRAW_DMA2D_MODE_MEMORY_TO_MEMORY_WITH_BLENDING_AND_FIXED_COLOR_BG
+} lv_draw_dma2d_mode_t;
+
+typedef struct {
+    lv_draw_dma2d_mode_t mode;
+    uint32_t w;
+    uint32_t h;
+
+    void * output_address;
+    uint32_t output_offset;
+    lv_draw_dma2d_output_cf_t output_cf;
+
+    uint32_t reg_to_mem_mode_color;
+
+    uint32_t fg_color;
+    uint32_t fg_alpha_mode; /* 3 possible modes - 0x0 to 0x2 */
+    uint32_t fg_alpha;
+
+    void * bg_address;
+    uint32_t bg_offset;
+    lv_draw_dma2d_fgbg_cf_t bg_cf;
+
+} lv_draw_dma2d_configuration_t;
+
+typedef struct {
+    void * first_byte;
+    uint32_t width_bytes;
+    uint32_t height;
+    uint32_t stride;
+} lv_draw_dma2d_cache_area_t;
+
+typedef struct {
+    lv_draw_unit_t base_unit;
+    lv_draw_task_t * task_act;
+    lv_draw_dma2d_cache_area_t writing_area;
+#if LV_USE_OS
+    lv_thread_t thread;
+#if LV_USE_DRAW_DMA2D_INTERRUPT
+    lv_thread_sync_t interrupt_signal;
+#else
+    lv_thread_sync_t thread_begin_polling_signal;
+#endif
+#endif
+} lv_draw_dma2d_unit_t;
+
+/**********************
+ *  STATIC PROTOTYPES
+ **********************/
+
+static int32_t evaluate_cb(lv_draw_unit_t * draw_unit, lv_draw_task_t * task);
+static int32_t dispatch_cb(lv_draw_unit_t * draw_unit, lv_layer_t * layer);
+static int32_t delete_cb(lv_draw_unit_t * draw_unit);
+static void opaque_fill(void * first_pixel, int32_t w, int32_t h, int32_t stride, lv_color_t color,
```

Comment:
Probably just personal preference, but I'd add `lv_draw_dma2d_unit_t * u` as the first parameter.